2177: Use tupaia service in Strive

kael89 opened the pull request:
### Issue #:
https://github.com/beyondessential/tupaia-backlog/issues/2177

This PR depends on the following features for correct functionality:
1. [1149: Correct internal data fetching time constraints to respect timezone](https://github.com/beyondessential/tupaia-backlog/issues/1149)
2. [2101: Analytics table](https://github.com/beyondessential/tupaia-backlog/issues/2101)
3. [2100: Use data_time field in analytics table](https://github.com/beyondessential/tupaia-backlog/issues/2100)
4. support for empty `dataElementCodes` when fetching tupaia events

All the above are included in the base branch. 1 and 2 are currently under testing. 3 and 4 depend on 1 and 2 so they will be reviewed and tested after their dependencies pass testing.

One can still review this PR, since none of its prerequisites affect the core migration logic.

kael89 commented on packages/database/src/migrations/20210318031303-UseTupaiaServiceInStrive-modifies-data.js:
```diff
@@ -0,0 +1,279 @@
+  seed = seedLink;
+};
+
+const ENTITY_AGGREGATION_CONFIG = {
```
I placed all "standard" tasks that need to be done when converting a project to `tupaia` service in this single migration file, to make future conversions easier

chris-bes commented on packages/database/src/migrations/20210318031303-UseTupaiaServiceInStrive-modifies-data.js:
```diff
@@ -0,0 +1,279 @@
+    STRIVE_WTF_mRDT_Tests_Radius: { dataSourceEntityType: 'facility' },
+    STRIVE_WTF_Positive: { dataSourceEntityType: 'facility' },
+  },
+};
```
I didn't really go through this list to check of the `dataSourceEntityType` or `aggregationEntityType` are correct, but I assume you've already smoke tested them 👍

kael89 replied:
Good call I believe, yes I've smoke tested them, and we'll do some thorough regression testing which will hopefully catch anything that may be wrong!

chris-bes reviewed (approved):
LGTM!

kael89 commented:
@chris-bes I added a [new commit](https://github.com/beyondessential/tupaia/pull/2287/commits/7c13941881a53255aee7ee2586a2feb57ab89bd3), could you please re-review?

There are branches for all the dependencies mentioned in this PR's description now, so I rebased the branches in this PR accordingly.

chris-bes commented on packages/database/src/migrations/20210322030033-FixStriveAfterTupaiaServiceConversion-modifies-data.js:
```diff
@@ -14,7 +14,17 @@ exports.setup = function (options, seedLink) {
+    WHERE
+      sr.entity_id = ${table}.record_id AND
+      s.code = 'SCRF';
+`);
```
Sorry if I'm being a bit pedantic, but I think not all survey responses are attached to case entity in `SCRF` survey (not sure why, but I think it could be a mistake somehow)...? 

When I run this query (converted from your delete statement):
```
SELECT distinct e.type from dhis_sync_log d
JOIN survey_response sr on sr.entity_id = d.record_id
JOIN survey s on s.id = sr.survey_id
JOIN entity e on e.id = sr.entity_id
WHERE s.code = 'SCRF'
```

It gives the result (expected only case, but not sure how a facility slipped through):
![image](https://user-images.githubusercontent.com/63621318/113954733-c0dcac00-985d-11eb-85b7-bb1d4527cac6.png)

Tracing the facility entity sync records, it is only 1 facility:
![facility](https://user-images.githubusercontent.com/63621318/113954889-131dcd00-985e-11eb-94ac-5e705bf2794d.PNG)

So I guess we don't want to delete the sync record of the facility entities If I'm correct... If so we can add a where clause `e.type = 'case'` to the delete statement?

kael89 replied:
Not pedantic at all, on the contrary thanks for giving this a thorough look! Yeah not sure why a `facility` slipped through, but it's good to keep it safe and add the extra condition you mentioned

chris-bes reviewed (approved):
Looks good!

kael89 commented:
Merging in the base branch so that  [2636](https://github.com/beyondessential/tupaia-backlog/issues/2636) and [2177](https://github.com/beyondessential/tupaia-backlog/issues/2177) can be tested together
